src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -53,12 +57,169 @@ bool CDROM_Interface_Ioctl::GetUPC(unsigned char &attr, char *upc)
+#endif
+		return false;
+	}
+	stTrack     = toc.cdth_trk0;
+	end         = toc.cdth_trk1;
+	leadOut.min = entry.cdte_addr.msf.minute;
+	leadOut.sec = entry.cdte_addr.msf.second;
+	leadOut.fr  = entry.cdte_addr.msf.frame;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: GetAudioTracks => start track is %2d, last playable track is %2d, "
+	         "and lead-out MSF is %02d:%02d:%02d",
+	         stTrack,
+	         end,
+	         leadOut.min,
+	         leadOut.sec,
+	         leadOut.fr);
+#endif
+	return true;
+}
+
+bool CDROM_Interface_Ioctl::GetAudioTrackInfo(uint8_t track, TMSF& start,
+                                              unsigned char& attr)
+{
+	if (cdrom_fd == -1) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioTrackInfo: cdrom_fd not open");
+#endif
+		return false;
+	}
+	cdrom_tocentry entry = {};
+	entry.cdte_track     = track;
+	entry.cdte_format    = CDROM_MSF;
+	if (ioctl(cdrom_fd, CDROMREADTOCENTRY, &entry) != 0) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioTrackInfo: ioctl CDROMREADTOCENTRY failed");
+#endif
+		return false;
+	}
+	start.min = entry.cdte_addr.msf.minute;
+	start.sec = entry.cdte_addr.msf.second;
+	start.fr  = entry.cdte_addr.msf.frame;
+	attr      = (entry.cdte_ctrl << 4) | entry.cdte_adr;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: GetAudioTrackInfo for track %u => "
+	         "MSF %02d:%02d:%02d, which is sector %d",
+	         track,
+	         start.min,
+	         start.sec,
+	         start.fr,
+	         msf_to_frames(start));
+#endif
+	return true;
+}
+
+bool CDROM_Interface_Ioctl::GetAudioSub(unsigned char& attr, unsigned char& track,
+                                        unsigned char& index, TMSF& relPos,
+                                        TMSF& absPos)
+{
+	if (cdrom_fd == -1) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioSub: cdrom_fd not open");
+#endif
+		return false;
+	}
+	cdrom_subchnl sub = {};
+	sub.cdsc_format   = CDROM_MSF;
+	if (ioctl(cdrom_fd, CDROMSUBCHNL, &sub) != 0) {
+#ifdef DEBUG_IOCTL
+		LOG_WARNING("CDROM_IOCTL: GetAudioSub: ioctl CDROMSUBCHNL failed");
+#endif
+		return false;
+	}
+	attr       = (sub.cdsc_ctrl << 4) | sub.cdsc_adr;
+	track      = sub.cdsc_trk;
+	index      = sub.cdsc_ind;
+	relPos.min = sub.cdsc_reladdr.msf.minute;
+	relPos.sec = sub.cdsc_reladdr.msf.second;
+	relPos.fr  = sub.cdsc_reladdr.msf.frame;
+	absPos.min = sub.cdsc_absaddr.msf.minute;
+	absPos.sec = sub.cdsc_absaddr.msf.second;
+	absPos.fr  = sub.cdsc_absaddr.msf.frame;
+#ifdef DEBUG_IOCTL
+	LOG_INFO("CDROM_IOCTL: GetAudioSub => position at %02d:%02d:%02d (on sector %u) "
+	         "within track %u at %02d:%02d:%02d (at its sector %u)",
+	         absPos.min,
+	         absPos.sec,
+	         absPos.fr,
+	         msf_to_frames(absPos),
+	         track,
+	         relPos.min,
+	         relPos.sec,
+	         relPos.fr,
+	         msf_to_frames(relPos));
+#endif
+	return true;
+}
```

Comment:
Same general comment: please insert some blank lines to group things logically and break up the wall-of-text look. This is hard to parse visually. I'd insert blanks around the `if` and before the last debug log statement.

Reply:
Fixed